src/Microsoft.NetCore.Analyzers/Core/Tasks/UseValueTasksCorrectly.cs

```diff
@@ -0,0 +1,547 @@
+                {
+                    return;
+                }
+
+                // Get the type for System.Diagnostics.Debug.  If we can't find it, that's ok, we just won't use it.
+                compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemDiagnosticsDebug, out var debugType);
+
+                // Process all invocations.  This analyzer works by finding all method invocations that return ValueTasks,
+                // and then analyzing those invocations and their surroundings.
+                compilationContext.RegisterOperationAction(operationContext =>
+                {
+                    // Is this a method invocation?  If not, we're done.
+                    var operation = operationContext.Operation;
+                    var invocation = (IInvocationOperation)operation;
+                    if (invocation.TargetMethod is null ||
+                        invocation.TargetMethod.MethodKind == MethodKind.Constructor) // ignore 'new ValueTask(...)'
+                    {
+                        return;
+                    }
+
+                    // Does the method return ValueTask?  If not, we're done.
+                    if (!valueTaskType.Equals(invocation.TargetMethod.ReturnType) &&
+                        !valueTaskOfTType.Equals(invocation.TargetMethod.ReturnType.OriginalDefinition))
+                    {
+                        return;
+                    }
+
+                    // The only method that returns a ValueTask on which we allow unlimited later consumption
+                    // is ValueTask.Preserve.  Use is rare, but special-case it.  If this is Preserve, we're done.
+                    if (invocation.TargetMethod.Name == "Preserve" &&
+                        (valueTaskType.Equals(invocation.TargetMethod.ContainingType) ||
+                         valueTaskOfTType.Equals(invocation.TargetMethod.ContainingType.OriginalDefinition)))
+                    {
+                        return;
+                    }
+
+                    // If this is a method call off of the ValueTask, then check to see if it's special.
+                    if (invocation.Parent is IInvocationOperation parentIo && parentIo.TargetMethod != null)
+                    {
+                        switch (parentIo.TargetMethod.Name)
+                        {
+                            case "AsTask":
+                            case "Preserve":
+                                // Using AsTask to convert to a Task is acceptable consumption.
+                                // Use Preserve to enable subsequent unlimited consumption is acceptable.
+                                return;
+
+                            case "GetAwaiter" when parentIo.Parent is IInvocationOperation ga && ga.TargetMethod != null && ga.TargetMethod.Name == "GetResult":
```

Comment:
`when parentIo.Parent is IInvocationOperation { TargetMethod: { Name: "GetResult"} }`
I have never seen nested pattern matching actually used in production code, and here is an excellent case.

Reply:
Lots of opportunities for this.